### ISSUE TITLE
rpn_cls_loss doesn't work too many wrong proposals with the rpn for gland data

### ISSUE THREAD

**Issue by lyn1874** [matterport/Mask_RCNN#443]
I want to make sure the RPN which is the first state of Mask RCNN work, so I only trained the RPN. I trained 40001 iterations, initial lr is 0.001, backbone is ResNet_V2_50. But I got several problems:

1. I found out the weight basically is not updated after 5000 iterations, and the gradients are really close to zero. wired weight
![9d0f74b3-eda0-4db3-a810-63ca2a8f1a59](https://user-images.githubusercontent.com/22235764/38741928-72e9b4f8-3f3b-11e8-8ef5-5189d3f3f825.jpg)
 I have tried to use a lower lr, and the epsilon I use in the tf.train.AdamOptimizer is 0.0001. Still the weight seems not updated.

2. As for the prediction, I feel the bounding box prediction is quite ok 
![pred](https://user-images.githubusercontent.com/22235764/38741984-8f931824-3f3b-11e8-844e-c28283898891.png)
 (green are the ground truth, red are the predicted box, these predicted boxes are selected if the anchors has a overlapping ratio larger than 0.7 with the ground truth bounding box), predicted bounding box, so it means at least the bounding box regression loss can make the prediction close to the ground truth. But as for the rpn_cls_loss, it classify too many wrong proposals to be object 
![figure_2](https://user-images.githubusercontent.com/22235764/38741990-950a2888-3f3b-11e8-8952-64c8c44ec3bb.png)
. I have consider the class imbalance problem(in gland cell dataset, average number of positive anchors is only 20 per image, only 85 training images), so I random choose 128 positive anchor and 128 negative anchors (if the positive anchor smaller than 128, I random_shuffle the index so one positive anchor will appear several times), but still it doesn't work. One thing I don't understand is that when we calculate the loss, we only consider a minibatch, but at the same time, we classify all the anchors that we generated to be a object or non-object, so what about those anchors which are not contributing to the loss? Can the network also make correct decision for those anchors?
I would really appreciate your help!! Thanksss  a lot!!!!!!

**Comment by lyn1874** [comment]
@waleedka Thank you very much!!!!!

**Comment by lyn1874** [comment]
@waleedka I calculate the rpn recall as in the inspect_model file, and I found out if I don't do the NMS, only clip the boxes w.r.t image coordinate, the precision could be 0.89 for my test images(but there are around 100000 boxes). But if I use NMS, the recall precision is decreased to be 0.6 (The proposal number is 50000, and threshold is 0.8), the reason for this is because the score for boxes belongs to  a object is quite unique, e.g. the only probability for boxes belong to an object is 0.68 or in some other image it will be 0.8, which means perform NMS for these boxes based on the score will not really select those correct boxes(because class prediction is not really accurate). I think maybe it would be better for me to not use NMS for the RPN stage, so that the second stage has more chance to learn the correct boxes?

**Comment by lyn1874** [comment]
Thanks a lot. It works very well. The recall for the boxes after clipped based on windows are 0.91. Then the recall for the boxes after NMS (propose number is 2000) is 0.89.

**Comment by xelmirage** [comment]
@lyn1874 Hi，I'm facing a similar problem with you and want to do some troubleshooting, but my tensorboard didn't give me the gradient like yours, I'm wondering did you made some modification on the code to show the gradient? Or it is related with tensorflow/keras/tensorboard version?
Thanks
